pkg/core/native/oracle.go

```diff
@@ -0,0 +1,432 @@
+	reqItem := &state.StorageItem{Value: req.Bytes()}
+	reqKey := makeRequestKey(id)
+	if err := d.PutStorageItem(o.ContractID, reqKey, reqItem); err != nil {
+		return err
+	}
+
+	// Add request ID to the id list.
+	lst := new(IDList)
+	key := makeIDListKey(req.URL)
+	if err := o.getSerializableFromDAO(d, key, lst); err != nil && !errors.Is(err, storage.ErrKeyNotFound) {
+		return err
+	}
+	*lst = append(*lst, id)
+	si := &state.StorageItem{Value: lst.Bytes()}
+	return d.PutStorageItem(o.ContractID, key, si)
+}
+
+func (o *Oracle) getOracleNodes(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	pubs := o.GetOracleNodes()
+	return pubsToArray(pubs)
+}
+
+// GetOracleNodes returns public keys of oracle nodes.
+func (o *Oracle) GetOracleNodes() keys.PublicKeys {
+	return o.nodes.Load().(keys.PublicKeys).Copy()
+}
+
+// GetScriptHash returns script hash or oracle nodes.
+func (o *Oracle) GetScriptHash() (util.Uint160, error) {
+	h := o.oracleHash.Load()
+	if h == nil {
+		return util.Uint160{}, storage.ErrKeyNotFound
+	}
+	return h.(util.Uint160), nil
+}
+
+func (o *Oracle) setOracleNodes(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	var pubs keys.PublicKeys
+	err := o.SetOracleNodes(ic, pubs)
+	if err != nil {
+		panic(err)
+	}
+	return pubsToArray(pubs)
+}
+
+// SetOracleNodes sets oracle node public keys to pubs.
+func (o *Oracle) SetOracleNodes(ic *interop.Context, pubs keys.PublicKeys) error {
+	if len(pubs) == 0 {
+		return ErrEmptyNodeList
+	}
+	h := o.NEO.GetCommitteeAddress()
+	if ok, err := runtime.CheckHashedWitness(ic, h); err != nil || !ok {
+		return ErrInvalidWitness
+	}
+
+	sort.Sort(pubs)
+	o.nodes.Store(pubs)
+	script, err := smartcontract.CreateMajorityMultiSigRedeemScript(pubs.Copy())
+	if err != nil {
+		return err
+	}
+	o.oracleHash.Store(hash.Hash160(script))
+	si := &state.StorageItem{Value: NodeList(pubs).Bytes()}
+	return ic.DAO.PutStorageItem(o.ContractID, prefixNodeList, si)
+}
+
+// GetRequestInternal returns request by ID and key under which it is stored.
+func (o *Oracle) GetRequestInternal(d dao.DAO, id uint64) (*OracleRequest, error) {
+	key := makeRequestKey(id)
+	req := new(OracleRequest)
+	return req, o.getSerializableFromDAO(d, key, req)
+}
+
+// GetIDListInternal returns request by ID and key under which it is stored.
+func (o *Oracle) GetIDListInternal(d dao.DAO, url string) (*IDList, error) {
+	key := makeIDListKey(url)
+	idList := new(IDList)
+	return idList, o.getSerializableFromDAO(d, key, idList)
+}
+
+func (o *Oracle) verify(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.NewBool(ic.Tx.HasAttribute(transaction.OracleResponseT))
+}
+
+func (o *Oracle) getOriginalTxID(d dao.DAO, tx *transaction.Transaction) util.Uint256 {
+	for i := range tx.Attributes {
+		if tx.Attributes[i].Type == transaction.OracleResponseT {
+			id := tx.Attributes[i].Value.(*transaction.OracleResponse).ID
+			req, _ := o.GetRequestInternal(d, id)
+			return req.OriginalTxID
```

Comment:
I'm not sure about this btw, it's not covered at the moment and it's a little strange to me that is only goes one level deep.

Reply:
What do you mean by "one level deep"?